videoanalyser/app/src/main/java/com/example/videoanalyser/MainActivity.java

```diff
@@ -0,0 +1,132 @@
+import android.widget.TextView;
+
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.mlkit.vision.common.InputImage;
+import com.google.mlkit.vision.text.Text;
+import com.google.mlkit.vision.text.TextRecognition;
+import com.google.mlkit.vision.text.TextRecognizer;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+
+import static android.media.MediaMetadataRetriever.METADATA_KEY_VIDEO_FRAME_COUNT;
+
+public class MainActivity extends AppCompatActivity {
+    private TextView filePath;
+    private Button filePickerBtn;
+    private Button analyzeBtn;
+    private Intent fileIntent;
+    private MediaMetadataRetriever mediaMetadataRetriever;
+    private Uri fileUri;
+    private  Integer totalFrames=0;
+    private  Integer safeFrames=10;
+    private String rotationOfVideo;
+    private InputImage imageHolder;
+    private TextRecognizer recognizer;
+    private ArrayList<String> results = new ArrayList<String>();
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        filePath = findViewById(R.id.selectedFilePath);
+        filePickerBtn = findViewById(R.id.fileSelectButton);
+        analyzeBtn = findViewById(R.id.analyze);
+        recognizer = TextRecognition.getClient();
+
+        filePickerBtn.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                fileIntent = new Intent(Intent.ACTION_GET_CONTENT);
+                fileIntent.setType("*/*");
+                startActivityForResult(fileIntent, 10);
+            }
+
+        });
+
+        analyzeBtn.setOnClickListener(new View.OnClickListener() {
+            @RequiresApi(api = Build.VERSION_CODES.P)
+            @Override
+            public void onClick(View view) {
+                mediaMetadataRetriever = new MediaMetadataRetriever();
+
+                try {
+                    mediaMetadataRetriever.setDataSource(getApplicationContext(), fileUri);
+                }catch (Exception e){
+                    e.printStackTrace();
+                    e.getCause();
+                }
+
+                totalFrames = Integer.valueOf(mediaMetadataRetriever.extractMetadata(METADATA_KEY_VIDEO_FRAME_COUNT));
+                List<Bitmap> frameList = mediaMetadataRetriever.getFramesAtIndex(0, Math.max(0, totalFrames-safeFrames));
+
+                for(int i=0; i<frameList.size();i++)
+                {
+                    // analysis will be done here (forking in a separate thread)
+                    imageHolder = InputImage.fromBitmap(frameList.get(i), 0);
+                    final int finalI = i;
+                    Task<Text> result =
```

Comment:
Observation: Right now this is an experiment, but when we get to implementing this for real, I think it makes sense to track/manage the number of asynchronous tasks you have going on at once to avoid exhausting system resources. This may occur naturally in a binary search implementation anyway, as the next frame to analyse depends on the result of the earlier frame.

Reply:
Intresitng idea! I will patch that

Reply:
We will also need a way to get the results back into the main thread.

Most likely have worker threads with a queue of tasks in one direction and results in the other direction.